src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/LanguageServiceProtocol/RedesignedLanguageServerTests/Nl2FxTests.cs

```diff
@@ -231,8 +232,35 @@ private static (string payload, string id) NL2FxMessageJson(string documentUri)
         private TestNL2FxHandler CreateAndConfigureNl2FxHandler()
         {
             var nl2FxHandler = new TestNL2FxHandler();
-            HandlerFactory.SetHandler(CustomProtocolNames.NL2FX, new Nl2FxLanguageServerOperationHandler(new BackwardsCompatibleNLHandlerFactory(nl2FxHandler)));
+            HandlerFactory.SetHandler(CustomProtocolNames.NL2FX, new Nl2FxLanguageServerOperationHandler(TestNlHandlerFactory.Create(nl2FxHandler)));
             return nl2FxHandler;
+        }
+
+        private sealed class TestNlHandlerFactory : BackwardsCompatibleNLHandlerFactory, IAsyncNLHandlerFactory
+        {
+            public int DelayTime { get; set; } = 50;
+
+            public TestNlHandlerFactory(NLHandler handler)
+                : base(handler)
+            {
+            }
+
+            public async Task<NLHandler> GetNLHandlerAsync(IPowerFxScope scope, BaseNLParams nlParams, CancellationToken cancellationToken = default)
+            {
+                if (DelayTime > 0)
+                {
+                    await Task.Delay(DelayTime, cancellationToken).ConfigureAwait(false);
+                }
+
+                return base.GetNLHandler(scope, nlParams);
+            }
+
+            public static INLHandlerFactory Create(NLHandler handler)
+            {
+                var random = new Random();
+                var useAsyncFactory = random.Next(0, 2) == 1;

```

Comment:
Is there a possibility (even small) that a test will pass with one case and not with the other? Having random values may lead to flaky tests. Can we pass the async value as a parameter, and in the tests we can be explicit about which one to use, possibly calling it with both values?

Reply:
I don't see a chance for flakiness here. while i agree that randomness should be avoided in tests, i tend to respect that as far as introducing randomness in the test cases or test inputs go (generating test cases randomly).  that said, i will remove it anyways